### PR TITLE
Resolve and Reject permissions promise on ios < 10

### PR DESCRIPTION
**Pull request by mcorner**
First stab at this, I would welcome comments.  The most unsatisfying part is rejecting the promise on any of the notification types being missing, but at least it will resolve.

Thanks for the great module!

### CODE REVIEW

File: ios/RNFIRMessaging.m [hunk 1]
```diff
@@ -247,12 +268,46 @@ - (void)messaging:(nonnull FIRMessaging *)messaging didRefreshRegistrationToken:
...
 RCT_EXPORT_METHOD(requestPermissions:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
 {
     if (RCTRunningInAppExtension()) {
         return;
     }
     if (floor(NSFoundationVersionNumber) <= NSFoundationVersionNumber_iOS_9_x_Max) {
+        // For ios <=9 we will use callbacks in AppDelegate.m to get notification of permission granted/denied
+        if ((_requestPermissionsResolveBlock != nil) || (_requestPermissionsRejectBlock != nil)) {
+          RCTLogError(@"Cannot call requestPermissions twice before the first has returned.");
+          return;
```
Comment (evollu) [inline ios/RNFIRMessaging.m]: can you call reject here as well? since JS world should know about the error

File: ios/RNFIRMessaging.m [hunk 2]
```diff
@@ -247,12 +268,46 @@ - (void)messaging:(nonnull FIRMessaging *)messaging didRefreshRegistrationToken:
...
 
+- (void)handleRegisterUserNotificationSettings:(NSNotification *)notification
+{
+    if (_requestPermissionsResolveBlock == nil) {
+      return;
+    }
+
+    UIUserNotificationSettings *notificationSettings = notification.userInfo[@"notificationSettings"];
+
+    if (((notificationSettings.types & UIUserNotificationTypeAlert) == 0) ||
```
Comment (evollu) [inline ios/RNFIRMessaging.m]: can we only reject if all of the permissions are declined?

File: ios/RNFIRMessaging.m [hunk 3]
```diff
@@ -170,6 +176,16 @@ + (void)willPresentNotification:(UNNotification *)notification withCompletionHan
     [[NSNotificationCenter defaultCenter] postNotificationName:FCMNotificationReceived object:self userInfo:@{@"data": data, @"completionHandler": completionHandler}];
 }
 
++ (void)didRegisterUserNotificationSettings:(__unused UIUserNotificationSettings *)notificationSettings
```
Comment (evollu) [inline ios/RNFIRMessaging.m]: can this method be combined with `handleRegisterUserNotificationSettings`?

Comment (mcorner) [inline ios/RNFIRMessaging.m]: Not easily.  This connects the class method to the instance method via broadcast.  One could convert to a singleton perhaps?  And this fits with how other things are done here. Am I missing something?

File: ios/RNFIRMessaging.m (hunk 2 above)
Comment (mcorner) [inline ios/RNFIRMessaging.m]: I think there is a better way to do this in general, which is to pass what notifications you want to requestPermissions and then resolve/reject based on those.  I will try and cook that up now.

File: ios/RNFIRMessaging.m (hunk 1 above)
Comment (mcorner) [inline ios/RNFIRMessaging.m]: agreed.

**Comment by mcorner** [comment]
I took a stab at asking for specific kinds of alerts and then rejecting if you don't get what you want.  You have to make a second call in ios10 to get the permissions (this took a bit of figuring out).  I am not 100% on the alertStyle in ios10, but this was my best guess.  Let me know how this looks.

**Comment by evollu** [comment]
specific notification permisson sound like a minor use case.
I'm travelling now, will test and merge this Sunday

**Comment by evollu** [comment]
closing due to inactivity

**Pull request closed without merging**